File: java/code/binary_search_step.java
import java.io.*;
import java.util.*;
import javax.swing.*;
import javax.swing.filechooser.FileNameExtensionFilter;

import javax.swing.JFileChooser;

public class binary_search_step {

    static class Pair {
        long number;
        String text;

        Pair(long number, String text) {
            this.number = number;
            this.text = text;
        }
    }

    public static void main(String[] args) {
        String fileName = GetUserInput();
        boolean fileIsFound = SearchFileName(fileName);

        File file = new File(fileName);
        String baseName = file.getName();

        if (!fileIsFound) {
            System.out.println(baseName + " was NOT FOUND at the given path.");
            return;
        }

        System.out.println(baseName + " was FOUND.");

        Scanner targetObject = new Scanner(System.in);
        System.out.println("Enter a number to search:");
        String targetInput = targetObject.nextLine();

        List<Pair> dataList = ParseFile(fileName);
        if (dataList == null) {
            System.out.println("Failed to read or parse the file.");
            return;
        }

        try {
            long target = Long.parseLong(targetInput);
            List<String> output = new ArrayList<>();
            int index = binarySearch(dataList, target, output);

            String baseDir = System.getProperty("user.dir");
            String folderPath = baseDir + File.separator + "java" + File.separator + "output" + File.separator
                    + "binary_search_step";
            File outputDir = new File(folderPath);
            if (!outputDir.exists()) {
                outputDir.mkdirs();
            }

            String status = (index != -1) ? "found" : "not found";
            String outputFilePath = folderPath + "/binary_search_step_" + target + ".txt (" + status + ")";
            try (PrintWriter writer = new PrintWriter(new FileWriter(outputFilePath))) {
                for (String step : output) {
                    writer.println(step);
                }
            }

            System.out.println("Search steps written to: " + outputFilePath);
        } catch (NumberFormatException e) {
            System.out.println("Invalid number entered. Please enter a valid number.");
        } catch (IOException e) {
            System.out.println("Error writing output file.");
        }
    }

    private static String GetUserInput() {
        JFileChooser fileChooser = new JFileChooser();
        fileChooser.setDialogTitle("Select a CSV dataset file for binary search step");
        fileChooser.setFileFilter(new javax.swing.filechooser.FileNameExtensionFilter("CSV files", "csv"));
        int result = fileChooser.showOpenDialog(null);

        if (result == JFileChooser.APPROVE_OPTION) {
            return fileChooser.getSelectedFile().getAbsolutePath();
        } else {
            System.out.println("No file selected. Exiting.");
            System.exit(0);
            return null; // unreachable, but required
        }
    }

    private static boolean SearchFileName(String fullPath) {
        File file = new File(fullPath);
        return file.exists() && file.isFile();
    }

    private static List<Pair> ParseFile(String filename) {
        List<Pair> list = new ArrayList<>();
        try (BufferedReader br = new BufferedReader(new FileReader(filename))) {
            String line;
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(",", 2);
                if (parts.length < 2) {
                    System.out.println("Invalid line format: " + line);
                    continue;
                }

                long number = Long.parseLong(parts[0].trim());
                String text = parts[1].trim();
                list.add(new Pair(number, text));
            }
        } catch (IOException | NumberFormatException e) {
            System.out.println("Error reading file: " + filename);
            return null;
        }
        return list;
    }

    private static int binarySearch(List<Pair> list, long target, List<String> stepsOutput) {
        int left = 0;
        int right = list.size() - 1;

        while (left <= right) {
            int middle = (left + right) / 2;
            long current = list.get(middle).number;

            String step = (middle + 1) + ": " + list.get(middle).number + "/" + list.get(middle).text;
            stepsOutput.add(step);

            if (current == target) {
                return middle;
            } else if (current > target) {
                right = middle - 1;
            } else {
                left = middle + 1;
            }
        }

        stepsOutput.add("-1");
        return -1;
    }
}

File: python/code/binary_search_step.py
import os
from typing import List, Optional, NamedTuple
import tkinter as tk
from tkinter import filedialog

class Pair(NamedTuple):
    number: int
    text: str

def select_file_with_dialog() -> Optional[str]:
    root = tk.Tk()
    root.withdraw()  # Hide the root window
    file_path = filedialog.askopenfilename(
        title="Select dataset file",
        filetypes=[("CSV Files", "*.csv"), ("Text Files", "*.txt"), ("All Files", "*.*")]
    )
    return file_path if file_path else None

def parse_file(filepath: str) -> Optional[List[Pair]]:
    """Parse the file and return a list of Pair objects."""
    data_list = []
    try:
        with open(filepath, 'r') as file:
            for line in file:
                line = line.strip()
                if line:
                    parts = line.split(",", 1)
                    if len(parts) == 2:
                        number = int(parts[0].strip())
                        text = parts[1].strip()
                        data_list.append(Pair(number, text))
        return data_list
    except (IOError, ValueError):
        print(f"Error reading file: {filepath}")
        return None

def binary_search(data_list: List[Pair], target: int, steps: List[str]) -> int:
    left, right = 0, len(data_list) - 1
    while left <= right:
        middle = (left + right) // 2
        current = data_list[middle].number
        steps.append(f"{middle + 1}: {data_list[middle].number}/{data_list[middle].text}")

        if current == target:
            return middle
        elif current > target:
            right = middle - 1
        else:
            left = middle + 1
    steps.append("-1")
    return -1

def write_steps_to_file(steps: List[str], target: int, found_index: int):
    base_dir = os.getcwd()
    python_output_dir = os.path.join(base_dir, "python", "output", "binary_search_step")
    os.makedirs(python_output_dir, exist_ok=True)

    result_suffix = "(target found)" if found_index != -1 else "(target not found)"
    output_filename = f"binary_search_step_{target}.txt {result_suffix}"
    output_path = os.path.join(python_output_dir, output_filename)

    with open(output_path, "w") as f:
        for step in steps:
            f.write(step + "\n")

    print(f"Steps written to: {output_path}")




def main():
    filepath = select_file_with_dialog()
    if not filepath:
        print("No file selected. Exiting.")
        return

    data_list = parse_file(filepath)
    if not data_list:
        print("Failed to read or parse the file.")
        return

    try:
        target_input = input("Enter a number to search: ")
        target = int(target_input)
    except ValueError:
        print("Invalid input. Please enter an integer.")
        return

    steps = []
    found_index = binary_search(data_list, target, steps)
    write_steps_to_file(steps, target, found_index)

if __name__ == "__main__":
    main()
